src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -1268,4 +1274,99 @@ ss::future<bool> tx_gateway_frontend::try_create_tx_topic() {
       });
 }
 
+void tx_gateway_frontend::expire_old_txs() {
+    (void)ss::with_gate(_gate, [this] {
+        auto shard = _shard_table.local().shard_for(model::tx_manager_ntp);
+
+        if (shard == std::nullopt) {
+            rearm_expire_timer();
+            return ss::now();
+        }
+
+        return container()
+          .invoke_on(
+            *shard,
+            _ssg,
+            [](tx_gateway_frontend& self) {
+                auto partition = self._partition_manager.local().get(
+                  model::tx_manager_ntp);
+                if (!partition) {
+                    vlog(
+                      clusterlog.warn,
+                      "can't get partition by {} ntp",
+                      model::tx_manager_ntp);
+                    return ss::now();
+                }
+
+                auto stm = partition->tm_stm();
+                return self.expire_old_txs(stm);
+            })
+          .finally([this] { rearm_expire_timer(); });
+    });
+}
```

Comment:
That's right there is a high probability that the GC phase is executed on a different core than a timer is working.

It'd be nice to run the loop on the right core and avoid the cross core communication but there're several problems with this approach:

  - The right core isn't known at the moment tx_gateway_frontend starts
  - Once we shard the transaction coordinator there may be several right
    cores

I could start the loop lazily but I don't think it's worth the work because the GC phase is very infrequent so the sub-optimality isn't a problem.

Reply:
I agree. did you document this in the code? the next person to read it will have the same question.